File: packages/room_scanner/android/src/main/kotlin/one/vron/room_scanner/RoomScannerPlugin.kt
package one.vron.room_scanner

import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.ar.core.ArCoreApk
import com.google.ar.core.Config
import com.google.ar.core.Session
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry

/**
 * Android implementation of room_scanner platform channel
 *
 * Uses ARCore Depth API for room scanning (Android 10+)
 *
 * Capabilities:
 * - ARCore depth API availability check
 * - Room scanning with real-time progress
 * - OBJ/PLY file export with depth data
 *
 * Note: LiDAR-quality scanning requires ARCore Depth API support
 */
class RoomScannerPlugin : FlutterPlugin, MethodCallHandler, ActivityAware,
    EventChannel.StreamHandler, PluginRegistry.RequestPermissionsResultListener {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    private var context: Context? = null
    private var activity: Activity? = null
    private var arSession: Session? = null
    private var currentSessionId: String? = null

    companion object {
        private const val METHOD_CHANNEL_NAME = "one.vron.mobile/room_scanner"
        private const val EVENT_CHANNEL_NAME = "one.vron.mobile/room_scanner_events"
        private const val CAMERA_PERMISSION_REQUEST_CODE = 1001
        private var pendingPermissionResult: Result? = null
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext

        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, METHOD_CHANNEL_NAME)
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, EVENT_CHANNEL_NAME)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        context = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // MARK: - MethodCallHandler

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "isLidarAvailable" -> handleIsLidarAvailable(result)
            "getDeviceCapabilities" -> handleGetDeviceCapabilities(result)
            "startScanning" -> handleStartScanning(call, result)
            "stopScanning" -> handleStopScanning(result)
            "cancelScanning" -> handleCancelScanning(result)
            "requestCameraPermission" -> handleRequestCameraPermission(result)
            else -> result.notImplemented()
        }
    }

    // MARK: - Method Handlers

    private fun handleIsLidarAvailable(result: Result) {
        val ctx = context
        if (ctx == null) {
            result.error("NO_CONTEXT", "Context not available", null)
            return
        }

        try {
            // Check if ARCore is installed and supported
            val arCoreAvailability = ArCoreApk.getInstance().checkAvailability(ctx)
            val isSupported = arCoreAvailability.isSupported

            // Check if device supports depth API
            val supportsDepth = if (isSupported) {
                checkDepthAPISupport(ctx)
            } else {
                false
            }

            result.success(supportsDepth)
        } catch (e: Exception) {
            result.error("CHECK_FAILED", "Failed to check LiDAR availability: ${e.message}", null)
        }
    }

    private fun handleGetDeviceCapabilities(result: Result) {
        val ctx = context
        if (ctx == null) {
            result.error("NO_CONTEXT", "Context not available", null)
            return
        }

        try {
            val arCoreAvailability = ArCoreApk.getInstance().checkAvailability(ctx)
            val hasDepthAPI = checkDepthAPISupport(ctx)

            val capabilities = mapOf(
                "hasLidar" to hasDepthAPI,
                "hasDepthAPI" to hasDepthAPI,
                "arCoreVersion" to getARCoreVersion(),
                "roomPlanSupported" to false, // Android uses different approach
                "minAndroidVersion" to "10.0",
                "supportedDevices" to listOf(
                    "Pixel 4 and later (with ARCore Depth API)",
                    "Samsung Galaxy S20+ and later (with ARCore Depth API)",
                    "Devices with ToF sensor or ARCore Depth API support"
                )
            )

            result.success(capabilities)
        } catch (e: Exception) {
            result.error("CAPABILITIES_FAILED", "Failed to get capabilities: ${e.message}", null)
        }
    }

    private fun handleStartScanning(call: MethodCall, result: Result) {
        val ctx = context
        val act = activity

        if (ctx == null || act == null) {
            result.error("NO_CONTEXT", "Context or Activity not available", null)
            return
        }

        val roomId = call.argument<String>("roomId")
        val roomName = call.argument<String>("roomName")
        val outputPath = call.argument<String>("outputPath")

        if (roomId == null || roomName == null || outputPath == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required arguments: roomId, roomName, outputPath",
                null
            )
            return
        }

        // Check depth API availability
        if (!checkDepthAPISupport(ctx)) {
            result.error(
                "LIDAR_NOT_AVAILABLE",
                "ARCore Depth API is not supported on this device",
                null
            )
            return
        }

        // Check camera permission
        if (ContextCompat.checkSelfPermission(ctx, android.Manifest.permission.CAMERA)
            != PackageManager.PERMISSION_GRANTED
        ) {
            result.error(
                "PERMISSION_DENIED",
                "Camera permission not granted",
                null
            )
            return
        }

        // Create new session
        val sessionId = java.util.UUID.randomUUID().toString()
        currentSessionId = sessionId

        // Initialize ARCore session (actual implementation would be more complex)
        // This is a placeholder for the native implementation
        try {
            arSession = Session(ctx)
            val config = Config(arSession)
            config.depthMode = Config.DepthMode.AUTOMATIC
            arSession?.configure(config)

            // Send initial progress
            eventSink?.success(
                mapOf(
                    "percentage" to 0.0,
                    "message" to "Starting room scan...",
                    "pointCount" to 0
                )
            )

            result.success(sessionId)
        } catch (e: Exception) {
            result.error("START_FAILED", "Failed to start scanning: ${e.message}", null)
        }
    }

    private fun handleStopScanning(result: Result) {
        if (currentSessionId == null) {
            result.error(
                "NO_ACTIVE_SESSION",
                "No active scanning session",
                null
            )
            return
        }

        // Stop the session and export data
        // This is a placeholder - actual implementation would:
        // 1. Stop ARCore session
        // 2. Process captured depth data
        // 3. Export to OBJ/PLY format
        // 4. Return file path

        try {
            arSession?.close()
            arSession = null

            val outputPath = "/path/to/generated/file.obj" // Placeholder
            currentSessionId = null

            result.success(outputPath)
        } catch (e: Exception) {
            result.error("STOP_FAILED", "Failed to stop scanning: ${e.message}", null)
        }
    }

    private fun handleCancelScanning(result: Result) {
        if (currentSessionId == null) {
            result.error(
                "NO_ACTIVE_SESSION",
                "No active scanning session",
                null
            )
            return
        }

        // Cancel the session without saving
        try {
            arSession?.close()
            arSession = null
            currentSessionId = null

            result.success(null)
        } catch (e: Exception) {
            result.error("CANCEL_FAILED", "Failed to cancel scanning: ${e.message}", null)
        }
    }

    private fun handleRequestCameraPermission(result: Result) {
        val act = activity
        if (act == null) {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }

        if (ContextCompat.checkSelfPermission(act, android.Manifest.permission.CAMERA)
            == PackageManager.PERMISSION_GRANTED
        ) {
            result.success(true)
            return
        }

        pendingPermissionResult = result
        ActivityCompat.requestPermissions(
            act,
            arrayOf(android.Manifest.permission.CAMERA),
            CAMERA_PERMISSION_REQUEST_CODE
        )
    }

    // MARK: - Permission Result Handler

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ): Boolean {
        if (requestCode == CAMERA_PERMISSION_REQUEST_CODE) {
            val granted = grantResults.isNotEmpty() &&
                    grantResults[0] == PackageManager.PERMISSION_GRANTED
            pendingPermissionResult?.success(granted)
            pendingPermissionResult = null
            return true
        }
        return false
    }

    // MARK: - EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // MARK: - Helper Methods

    private fun checkDepthAPISupport(context: Context): Boolean {
        return try {
            val session = Session(context)
            val config = Config(session)

            // Check if depth mode is supported
            val supportsDepth = session.isDepthModeSupported(Config.DepthMode.AUTOMATIC)

            session.close()
            supportsDepth
        } catch (e: Exception) {
            false
        }
    }

    private fun getARCoreVersion(): String {
        return try {
            // Get ARCore version from package manager
            val ctx = context ?: return "Unknown"
            val packageInfo = ctx.packageManager.getPackageInfo(
                "com.google.ar.core",
                0
            )
            packageInfo.versionName
        } catch (e: Exception) {
            "Unknown"
        }
    }
}

File: packages/room_scanner/ios/Classes/RoomScannerPlugin.swift
import Flutter
import UIKit
import ARKit
import RoomPlan

/// iOS implementation of room_scanner platform channel
///
/// Uses ARKit + RoomPlan framework for LiDAR scanning (iOS 16+)
///
/// Capabilities:
/// - LiDAR availability check
/// - Room scanning with real-time progress
/// - USDZ file export with room geometry
@available(iOS 16.0, *)
public class RoomScannerPlugin: NSObject, FlutterPlugin {
    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?

    private var roomCaptureSession: RoomCaptureSession?
    private var currentSessionId: String?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "one.vron.mobile/room_scanner",
            binaryMessenger: registrar.messenger()
        )

        let eventChannel = FlutterEventChannel(
            name: "one.vron.mobile/room_scanner_events",
            binaryMessenger: registrar.messenger()
        )

        let instance = RoomScannerPlugin()
        instance.methodChannel = methodChannel
        instance.eventChannel = eventChannel

        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "isLidarAvailable":
            handleIsLidarAvailable(result: result)

        case "getDeviceCapabilities":
            handleGetDeviceCapabilities(result: result)

        case "startScanning":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(
                    code: "INVALID_ARGUMENTS",
                    message: "Missing required arguments",
                    details: nil
                ))
                return
            }
            handleStartScanning(args: args, result: result)

        case "stopScanning":
            handleStopScanning(result: result)

        case "cancelScanning":
            handleCancelScanning(result: result)

        case "requestCameraPermission":
            handleRequestCameraPermission(result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method Handlers

    private func handleIsLidarAvailable(result: @escaping FlutterResult) {
        let isSupported = ARWorldTrackingConfiguration.supportsFrameSemantics(.sceneDepth)
        result(isSupported)
    }

    private func handleGetDeviceCapabilities(result: @escaping FlutterResult) {
        let hasLidar = ARWorldTrackingConfiguration.supportsFrameSemantics(.sceneDepth)
        let roomPlanSupported = RoomCaptureSession.isSupported

        let capabilities: [String: Any] = [
            "hasLidar": hasLidar,
            "roomPlanSupported": roomPlanSupported,
            "minIOSVersion": "16.0",
            "supportedDevices": [
                "iPhone 12 Pro",
                "iPhone 12 Pro Max",
                "iPhone 13 Pro",
                "iPhone 13 Pro Max",
                "iPhone 14 Pro",
                "iPhone 14 Pro Max",
                "iPhone 15 Pro",
                "iPhone 15 Pro Max",
                "iPad Pro 11-inch (2nd generation and later)",
                "iPad Pro 12.9-inch (4th generation and later)"
            ]
        ]

        result(capabilities)
    }

    private func handleStartScanning(args: [String: Any], result: @escaping FlutterResult) {
        guard let roomId = args["roomId"] as? String,
              let roomName = args["roomName"] as? String,
              let outputPath = args["outputPath"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: roomId, roomName, outputPath",
                details: nil
            ))
            return
        }

        // Check LiDAR availability
        guard RoomCaptureSession.isSupported else {
            result(FlutterError(
                code: "LIDAR_NOT_AVAILABLE",
                message: "RoomPlan is not supported on this device",
                details: nil
            ))
            return
        }

        // Check camera permission
        let authStatus = AVCaptureDevice.authorizationStatus(for: .video)
        guard authStatus == .authorized else {
            result(FlutterError(
                code: "PERMISSION_DENIED",
                message: "Camera permission not granted",
                details: nil
            ))
            return
        }

        // Create new session
        let sessionId = UUID().uuidString
        currentSessionId = sessionId

        // Initialize RoomCaptureSession (actual implementation would be more complex)
        // This is a placeholder for the native implementation

        // Send initial progress
        eventSink?([
            "percentage": 0.0,
            "message": "Starting room scan...",
            "pointCount": 0
        ])

        result(sessionId)
    }

    private func handleStopScanning(result: @escaping FlutterResult) {
        guard currentSessionId != nil else {
            result(FlutterError(
                code: "NO_ACTIVE_SESSION",
                message: "No active scanning session",
                details: nil
            ))
            return
        }

        // Stop the session and export USDZ
        // This is a placeholder - actual implementation would:
        // 1. Stop RoomCaptureSession
        // 2. Process captured data
        // 3. Export to USDZ format
        // 4. Return file path

        let outputPath = "/path/to/generated/file.usdz" // Placeholder
        currentSessionId = nil

        result(outputPath)
    }

    private func handleCancelScanning(result: @escaping FlutterResult) {
        guard currentSessionId != nil else {
            result(FlutterError(
                code: "NO_ACTIVE_SESSION",
                message: "No active scanning session",
                details: nil
            ))
            return
        }

        // Cancel the session without saving
        currentSessionId = nil
        roomCaptureSession = nil

        result(nil)
    }

    private func handleRequestCameraPermission(result: @escaping FlutterResult) {
        AVCaptureDevice.requestAccess(for: .video) { granted in
            DispatchQueue.main.async {
                result(granted)
            }
        }
    }
}

// MARK: - FlutterStreamHandler

@available(iOS 16.0, *)
extension RoomScannerPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}
